app/src/main/java/org/mozilla/focus/biometrics/BiometricAuthenticationHandler.kt

```diff
@@ -0,0 +1,180 @@
+                    throw RuntimeException("Failed to get an instance of Cipher", err)
+                else -> throw err
+            }
+        }
+
+        if (initCipher(defaultCipher, DEFAULT_KEY_NAME)) {
+            cryptoObject = FingerprintManagerCompat.CryptoObject(defaultCipher)
+        }
+
+        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
+    }
+
+    private fun initCipher(cipher: Cipher, keyName: String): Boolean {
+        try {
+            keyStore?.load(null)
+            val key = keyStore?.getKey(keyName, null) as SecretKey
+            cipher.init(Cipher.ENCRYPT_MODE, key)
+            return true
+        } catch (err: Exception) {
+            when (err) {
+                is KeyPermanentlyInvalidatedException -> return false
+                is KeyStoreException,
+                is CertificateException,
+                is UnrecoverableKeyException,
+                is IOException,
+                is NoSuchAlgorithmException,
+                is InvalidKeyException -> throw RuntimeException("Failed to init cipher", err)
+                else -> throw err
+            }
+        }
+    }
+
+    private fun createKey(keyName: String, invalidatedByBiometricEnrollment: Boolean?) {
+        try {
+            keyStore?.load(null)
+
+            val builder = KeyGenParameterSpec.Builder(keyName,
+                    KeyProperties.PURPOSE_ENCRYPT).setBlockModes(KeyProperties.BLOCK_MODE_CBC)
+                    .setUserAuthenticationRequired(true)
+                    .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_PKCS7)
+
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+                builder.setInvalidatedByBiometricEnrollment(invalidatedByBiometricEnrollment!!)
+            }
+
+            keyGenerator?.init(builder.build())
+            keyGenerator?.generateKey()
+
+        } catch (err: Exception) {
+            when (err) {
+                is NoSuchAlgorithmException,
+                is InvalidAlgorithmParameterException,
+                is CertificateException,
+                is IOException -> throw RuntimeException(err)
+                else -> throw err
+            }
+        }
+    }
+
+    // Create the prompt and begin listening
+    private fun startListening(cryptoObject: FingerprintManagerCompat.CryptoObject) {
+        if (!Biometrics.hasFingerprintHardware(context)) return
+        if (biometricFragment == null) {
+            biometricFragment = BiometricAuthenticationDialogFragment()
+        }
+
+        cancellationSignal = CancellationSignal()
+        selfCancelled = false
+        fingerprintManager.authenticate(cryptoObject, 0, cancellationSignal, this, null)
+    }
+
+    fun stopListening() {
+        cancellationSignal?.let {
+            selfCancelled = true
+            it.cancel()
+            cancellationSignal = null
+        }
+    }
+
+    fun startAuthentication() {
+        val cryptoObject = cryptoObject
```

Comment:
What is this line doing? Why do we have to assign a new variable here?

Reply:
This is necessary because cryptoObject is mutable and could change after the null check. By creating a val of cryptoObject we can check if it's null and call the interior function accordingly! Is there a better way to do this?